blockchain.go

```diff
@@ -42,3 +46,82 @@ func GetSigndiceTransaction(
 	log.Debug().Msg(signedTx.String())
 	return tx.Pack(eos.CompressionNone)
 }
+
+func ValidateDepositTransaction(
+	tx *eos.SignedTransaction,
+	casinoName, platformName eos.AccountName,
+	platformPubKey ecc.PublicKey,
+	chainID eos.Checksum256) error {
+	if len(tx.Actions) != 2 {
+		return fmt.Errorf("invalid actions size")
+	}
+	transferAction := tx.Actions[0]
+	if err := ValidateTransferAction(transferAction, casinoName); err != nil {
+		return err
+	}
+
+	transferData, ok := transferAction.Data.(token.Transfer)
+	if !ok {
+		return fmt.Errorf("invalid transfer data")
+	}
+
+	if err := ValidateNewGameAction(tx.Actions[1], platformName, transferData.To); err != nil {
+		return err
+	}
+
+	pubKeys, err := tx.SignedByKeys(chainID)
+	log.Debug().Msgf("Deposit txn pubkeys: %v", pubKeys)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve public keys from deposit transaction")
+	}
+	if err := ValidateSignatures(pubKeys, platformPubKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func ValidateTransferAction(action *eos.Action, casinoName eos.AccountName) error {
+	if action.Account != eos.AN("eosio.token") {
+		return fmt.Errorf("invalid contract name in transfer action")
+	}
+	if action.Name != eos.ActN("transfer") {
+		return fmt.Errorf("invalid action name in transfer action")
+	}
+	if len(action.Authorization) != 1 {
+		return fmt.Errorf("invalid authorization size in transfer action")
+	}
+	if string(action.Authorization[0].Permission) != string(casinoName) {
+		return fmt.Errorf("invalid permission in transfer action")
+	}
+	return nil
+}
+
+func ValidateNewGameAction(action *eos.Action, platformName, gameName eos.AccountName) error {
+	if action.Account != gameName {
+		return fmt.Errorf("invalid contract name in newgame action")
+	}
+	if action.Name != eos.ActN("newgame") {
+		return fmt.Errorf("invalid action name in newgame action")
+	}
+	if len(action.Authorization) != 1 {
+		return fmt.Errorf("invalid authorization size in newgame action")
+	}
+	if action.Authorization[0].Actor != platformName {
+		return fmt.Errorf("invalid actor in newgame action")
+	}
+	if action.Authorization[0].Permission != eos.PN("gameaction") {
+		return fmt.Errorf("invalid permission name in newgame action")
+	}
+	return nil
+}
+
+func ValidateSignatures(pubKeys []ecc.PublicKey, platformPubKey ecc.PublicKey) error {
+	// sponsorship and platform signatures
+	if len(pubKeys) != 2 {
```

Comment:
there are can be up to 3 signatures (platform deposit, platform gameaction, sponsor[optionally])